src/MongoDB.Driver/Linq/Processors/ExpressionValidator.cs

```diff
@@ -0,0 +1,122 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace MongoDB.Driver.Linq.Processors
+{
+    /// <summary>
+    /// The expression visitor to validate a LINQ expression.
+    /// </summary>
+    internal sealed class ExpressionValidator : ExpressionVisitor
+    {
+        private readonly List<string> _topLevelInParameters = new List<string>();
+        private List<string> _inParametersScope = new List<string>();
```

Comment:
Can you describe the purpose of these two fields? And perhaps name them more descriptively?

Reply:
Should _inParametersScope be a Set<string>?

It would simplify adding new parameters to the set.